File: game-engine/tests/engine_utils.test.ts
import * as Engine from '../src/engine_utils';
import { Board, StrategoBoard } from '../src/board/board';
import { Case, CaseState, createEmpty, createUnreachable, create as createCase } from '../src/case';
import { InitGameError } from '../src/error/error';
import { PieceType, Color, create as createPiece } from '../src/piece/piece';


test('Game should be over - because blue has looses Flag', () => {
    let cases = emptyBoard()

    //Red
    cases[0][0] = createCase(CaseState.Full, 0, 0, createPiece(PieceType.Flag, Color.Red));
    cases[0][1] = createCase(CaseState.Full, 0, 1, createPiece(PieceType.General, Color.Red));

    //Blue
    cases[9][1] = createCase(CaseState.Full, 9, 1, createPiece(PieceType.General, Color.Blue));

    let board = new StrategoBoard(cases);

    let res = Engine.gameIsOver(board);
    expect(res.ok).toBe(true);
    expect(res.val).toBe(Color.Red);
});

test('Game should be over - because blue has looses moveable pieces', () => {
    let cases = emptyBoard()

    //Red
    cases[0][0] = createCase(CaseState.Full, 0, 0, createPiece(PieceType.Flag, Color.Red));
    cases[0][1] = createCase(CaseState.Full, 0, 1, createPiece(PieceType.General, Color.Red));

    //Blue
    cases[9][0] = createCase(CaseState.Full, 9, 0, createPiece(PieceType.Flag, Color.Blue));

    let board = new StrategoBoard(cases);

    let res = Engine.gameIsOver(board);
    expect(res.ok).toBe(true);
    expect(res.val).toBe(Color.Red);
});

test('Game should not be over', () => {
    let cases = emptyBoard()

    //Red
    cases[0][0] = createCase(CaseState.Full, 0, 0, createPiece(PieceType.Flag, Color.Red));
    cases[0][1] = createCase(CaseState.Full, 0, 1, createPiece(PieceType.General, Color.Red));

    //Blue
    cases[9][0] = createCase(CaseState.Full, 9, 0, createPiece(PieceType.Flag, Color.Blue));
    cases[9][1] = createCase(CaseState.Full, 9, 1, createPiece(PieceType.General, Color.Blue));

    let board = new StrategoBoard(cases);

    let res = Engine.gameIsOver(board);
    expect(res.err).toBe(true);
});

function emptyBoard(): Case[][] {
    let size = 10;
    let board = new Array(size);
    for (var i: number = 0; i < size; i++) {
        board[i] = new Array(size);
        for (var j: number = 0; j < size; j++) {
            board[i][j] = createEmpty(i, j);
        }
    }

    board[4][2] = createUnreachable(4, 2);
    board[4][3] = createUnreachable(4, 3);
    board[5][2] = createUnreachable(5, 2);
    board[5][3] = createUnreachable(5, 3);
    board[4][6] = createUnreachable(4, 6);
    board[4][7] = createUnreachable(4, 7);
    board[5][6] = createUnreachable(5, 6);
    board[5][7] = createUnreachable(7, 5);

    return board;
}

test('Should not verify board integrity cause to small', () => {
    let newBoard: Case[][] = [
        [createEmpty(0, 0), createEmpty(0, 1)],
        [createEmpty(1, 0), createUnreachable(1, 1)]
    ];
    let res = Engine.verifyBoardIntegrity(newBoard);

    expect(res.err).toBe(true);
    expect(res.val).toStrictEqual(new InitGameError("Board is not official, GO OUT OF THERE !!"));
});

test('Should not verify board integrity cause lakes are empty', () => {
    let newBoard: Case[][] = createStrategoBoard();

    newBoard[4][2] = createCase(CaseState.Full, 4, 2, { move: { min: 0, max: 1 }, rank: 1, color: Color.Blue });

    let res = Engine.verifyBoardIntegrity(newBoard);

    expect(res.err).toBe(true);
    expect(res.val).toStrictEqual(new InitGameError("You can not place pieces in lakes, please check again"));
});

test('Should check only 2 rows in middle are empty', () => {
    let newBoard: Case[][] = new Array(10);

    for (let i = 0; i < 10; i++) {
        newBoard[i] = new Array(10);
        for (let j = 0; j < 10; j++) {
            newBoard[i][j] = createCase(
                CaseState.Full,
                i,
                j,
                { move: { min: 0, max: 1 }, rank: 1, color: Color.Blue }
            );
        }
    }

    newBoard[4][2] = createUnreachable(4, 2);
    newBoard[4][3] = createUnreachable(4, 3);
    newBoard[5][2] = createUnreachable(5, 2);
    newBoard[5][3] = createUnreachable(5, 3);
    newBoard[4][6] = createUnreachable(4, 6);
    newBoard[4][7] = createUnreachable(4, 7);
    newBoard[5][6] = createUnreachable(5, 6);
    newBoard[5][7] = createUnreachable(7, 5);

    //console.log("2 rows\n" + board.display());
    let res = Engine.verifyBoardIntegrity(newBoard);

    expect(res.err).toBe(true);
    expect(res.val).toStrictEqual(new InitGameError("The 2 rows in the middle must be empty, :("));
});

test('Should check players have placed theirs pieces in the 4 rows', () => {
    let newBoard: Case[][] = new Array(10);

    for (let i = 0; i < 10; i++) {
        newBoard[i] = new Array(10);
        for (let j = 0; j < 10; j++) {
            if (i < 4) {
                newBoard[i][j] = createCase(
                    CaseState.Full,
                    i,
                    j,
                    { move: { min: 0, max: 1 }, rank: 1, color: Color.Blue }
                );
            } else if (i > 5) {
                newBoard[i][j] = createCase(
                    CaseState.Full,
                    i,
                    j,
                    { move: { min: 0, max: 1 }, rank: 1, color: Color.Red }
                );
            } else {
                newBoard[i][j] = createEmpty(i, j);
            }
        }
    }

    // Placing a piece to the bad place
    newBoard[0][4] = createCase(CaseState.Full, 0, 4, { move: { min: 0, max: 1 }, rank: 1, color: Color.Red });

    newBoard[4][2] = createUnreachable(4, 2);
    newBoard[4][3] = createUnreachable(4, 3);
    newBoard[5][2] = createUnreachable(5, 2);
    newBoard[5][3] = createUnreachable(5, 3);
    newBoard[4][6] = createUnreachable(4, 6);
    newBoard[4][7] = createUnreachable(4, 7);
    newBoard[5][6] = createUnreachable(5, 6);
    newBoard[5][7] = createUnreachable(7, 5);

    //console.log("4 rows\n" + board.display());
    let res = Engine.verifyBoardIntegrity(newBoard);


    expect(res.err).toBe(true);
    expect(res.val).toStrictEqual(new InitGameError("Players must place theirs pieces in the right place !"));

});

test('Should check players have the right pieces', () => {
    let cases = createStrategoBoard();
    let board = new StrategoBoard(cases);
    //console.log(board.display());
    let res = Engine.verifyBoardIntegrity(cases);


    expect(res.err).toBe(true);
    expect(res.val).toStrictEqual(new InitGameError("You need to start with the right pieces"));
});

//test('Should verify board integrity', () => {
//let res = Engine.verifyBoardIntegrity(createStrategoBoard());

//expect(res.ok).toBe(true);
//});

//Bad board
function createStrategoBoard(): Case[][] {
    let newBoard: Case[][] = new Array(10);

    for (let i = 0; i < 10; i++) {
        newBoard[i] = new Array(10);
        for (let j = 0; j < 10; j++) {
            if (i < 4) {
                newBoard[i][j] = createCase(
                    CaseState.Full,
                    i,
                    j,
                    { move: { min: 0, max: 1 }, rank: 1, color: Color.Blue }
                );
            } else if (i > 5) {
                newBoard[i][j] = createCase(
                    CaseState.Full,
                    i,
                    j,
                    { move: { min: 0, max: 1 }, rank: 1, color: Color.Red }
                );
            } else {
                newBoard[i][j] = createEmpty(i, j);
            }
        }
    }

    newBoard[4][2] = createUnreachable(4, 2);
    newBoard[4][3] = createUnreachable(4, 3);
    newBoard[5][2] = createUnreachable(5, 2);
    newBoard[5][3] = createUnreachable(5, 3);
    newBoard[4][6] = createUnreachable(4, 6);
    newBoard[4][7] = createUnreachable(4, 7);
    newBoard[5][6] = createUnreachable(5, 6);
    newBoard[5][7] = createUnreachable(7, 5);

    return newBoard;
}

File: game-engine/stratego-lib/tests/engine_utils_test.rs
#[cfg(test)]
mod engine_utils_tests {

    use stratego_lib::board::case::{self, Case, Coordinate};
    use stratego_lib::board::classic_board::StrategoBoard;
    use stratego_lib::board::piece::{Color, Piece, PieceType};
    use stratego_lib::board::Board;
    use stratego_lib::engine_utils;


    #[test]
    fn game_should_be_over_because_blue_has_looses_flag() {
        let mut cases = empty_board();

        //Red
        cases[0][0] = case::create_full_case(
            Coordinate::new(0, 0),
            Piece::new(PieceType::Flag, Color::Red),
        );
        cases[0][1] = case::create_full_case(
            Coordinate::new(0, 1),
            Piece::new(PieceType::General, Color::Red),
        );

        //Blue
        cases[9][1] = case::create_full_case(
            Coordinate::new(9, 1),
            Piece::new(PieceType::General, Color::Blue),
        );

        let res = engine_utils::game_is_over(&cases);
        match res {
            Some(val) => {
                assert_eq!(Color::Red, val);
            }
            None => panic!("Should not happen"),
        }
    }
    #[test]
    fn game_should_be_over_because_blue_has_looses_moveable_pieces() {
        let mut cases = empty_board();

        //Red
        cases[0][0] = case::create_full_case(
            Coordinate::new(0, 0),
            Piece::new(PieceType::Flag, Color::Red),
        );
        cases[0][1] = case::create_full_case(
            Coordinate::new(0, 1),
            Piece::new(PieceType::General, Color::Red),
        );

        //Blue
        cases[9][0] = case::create_full_case(
            Coordinate::new(9, 0),
            Piece::new(PieceType::Flag, Color::Blue),
        );

        let res = engine_utils::game_is_over(&cases);
        match res {
            Some(val) => {
                assert_eq!(Color::Red, val);
            }
            None => panic!("Should not happen"),
        }
    }

    #[test]
    fn game_should_not_be_over() {
        let mut cases = empty_board();

        //Red
        cases[0][0] = case::create_full_case(
            Coordinate::new(0, 0),
            Piece::new(PieceType::Flag, Color::Red),
        );
        cases[0][1] = case::create_full_case(
            Coordinate::new(0, 1),
            Piece::new(PieceType::General, Color::Red),
        );

        //Blue
        cases[9][0] = case::create_full_case(
            Coordinate::new(9, 0),
            Piece::new(PieceType::Flag, Color::Blue),
        );
        cases[9][1] = case::create_full_case(
            Coordinate::new(9, 1),
            Piece::new(PieceType::General, Color::Blue),
        );

        let res = engine_utils::game_is_over(&cases);
        match res {
            Some(_) => panic!("Should not happen"),
            None => {
                assert!(true);
            }
        }
    }

    #[test]
    fn should_not_verify_board_integrity_cause_to_small() {
        let new_board = vec![
            vec![
                case::create_empty_case(Coordinate::new(0, 0)),
                case::create_empty_case(Coordinate::new(0, 1)),
            ],
            vec![
                case::create_empty_case(Coordinate::new(1, 0)),
                case::create_empty_case(Coordinate::new(1, 1)),
            ],
        ];

        let res = engine_utils::verify_board_integrity(StrategoBoard::new(new_board));

        match res {
            Ok(_) => panic!("Should not happen"),
            Err(e) => {
                assert!(true);
                assert_eq!(
                    e.message(),
                    String::from("Board is not official, GO OUT OF THERE !!")
                );
            }
        }
    }

    #[test]
    fn should_not_verify_board_integrity_cause_lakes_are_not_empty() {
        let mut new_board = create_statego_board();

        new_board[4][2] = case::create_full_case(
            Coordinate::new(4, 2),
            Piece::new(PieceType::Spy, Color::Blue),
        );

        let b = StrategoBoard::new(new_board.clone());
        println!("{}", b.display());
        let res = engine_utils::verify_board_integrity(StrategoBoard::new(new_board));

        match res {
            Ok(_) => panic!("Should not happen"),
            Err(e) => {
                assert!(true);
                assert_eq!(
                    e.message(),
                    String::from("You can not place pieces in lakes, please check again")
                );
            }
        }
    }

    #[test]
    fn should_check_only_2_rows_in_middle_are_empty() {
        let mut new_board = Vec::with_capacity(10);
        for i in 0..10 {
            new_board.push(Vec::with_capacity(10));
            for j in 0..10 {
                new_board[i].push(case::create_full_case(
                    Coordinate::new(i as i16, j as i16),
                    Piece::new(PieceType::Spy, Color::Blue),
                ));
            }
        }

        new_board[4][2] = case::create_unreachable_case(Coordinate::new(4, 2));
        new_board[4][3] = case::create_unreachable_case(Coordinate::new(4, 3));
        new_board[5][2] = case::create_unreachable_case(Coordinate::new(5, 2));
        new_board[5][3] = case::create_unreachable_case(Coordinate::new(5, 3));
        new_board[4][6] = case::create_unreachable_case(Coordinate::new(4, 6));
        new_board[4][7] = case::create_unreachable_case(Coordinate::new(4, 7));
        new_board[5][6] = case::create_unreachable_case(Coordinate::new(5, 6));
        new_board[5][7] = case::create_unreachable_case(Coordinate::new(7, 5));

        let res = engine_utils::verify_board_integrity(StrategoBoard::new(new_board));
        match res {
            Ok(_) => panic!("Should not happen"),
            Err(e) => {
                assert!(true);
                assert_eq!(
                    e.message(),
                    String::from("The 2 rows in the middle must be empty, :(")
                );
            }
        }
    }

    #[test]
    fn should_check_players_have_placed_theirs_pieces_in_the_four_rows() {
        let mut new_board = create_statego_board();
        new_board[0][4] = case::create_full_case(
            Coordinate::new(0, 4),
            Piece::new(PieceType::Spy, Color::Red),
        );
        let res = engine_utils::verify_board_integrity(StrategoBoard::new(new_board));
        match res {
            Ok(_) => panic!("Should not happen"),
            Err(e) => {
                assert!(true);
                assert_eq!(
                    e.message(),
                    String::from("Players must place theirs pieces in the right place !")
                );
            }
        }
    }

    #[test]
    fn should_check_players_have_the_right_pieces() {
        let cases = create_statego_board();
        let res = engine_utils::verify_board_integrity(StrategoBoard::new(cases));

        match res {
            Ok(_) => panic!("Should not happen"),
            Err(e) => {
                assert!(true);
                assert_eq!(
                    e.message(),
                    String::from("You need to start with the right pieces")
                );
            }
        }
    }


    //Good board
    fn create_statego_board() -> Vec<Vec<Case>> {
        let mut new_board = Vec::with_capacity(10);
        for i in 0..10 {
            new_board.push(Vec::with_capacity(10));
            for j in 0..10 {
                if i < 4 { new_board[i].push(case::create_full_case(
                        Coordinate::new(i as i16, j as i16),
                        Piece::new(PieceType::Spy, Color::Blue),
                    ));
                } else if i > 5 {
                    new_board[i].push(case::create_full_case(
                        Coordinate::new(i as i16, j as i16),
                        Piece::new(PieceType::Spy, Color::Red),
                    ));
                } else {
                    new_board[i].push(case::create_empty_case(Coordinate::new(i as i16, j as i16)));
                }
            }
        }

        new_board[4][2] = case::create_unreachable_case(Coordinate::new(4, 2));
        new_board[4][3] = case::create_unreachable_case(Coordinate::new(4, 3));
        new_board[5][2] = case::create_unreachable_case(Coordinate::new(5, 2));
        new_board[5][3] = case::create_unreachable_case(Coordinate::new(5, 3));
        new_board[4][6] = case::create_unreachable_case(Coordinate::new(4, 6));
        new_board[4][7] = case::create_unreachable_case(Coordinate::new(4, 7));
        new_board[5][6] = case::create_unreachable_case(Coordinate::new(5, 6));
        new_board[5][7] = case::create_unreachable_case(Coordinate::new(5, 7));

        new_board
    }

    fn empty_board() -> Vec<Vec<Case>> {
        let size = 10;
        let mut board = Vec::with_capacity(size);
        for i in 0..size {
            board.push(Vec::with_capacity(size));
            for j in 0..size {
                board[i].push(case::create_empty_case(Coordinate::new(i as i16, j as i16)));
            }
        }

        board[4][2] = case::create_unreachable_case(Coordinate::new(4, 2));
        board[4][3] = case::create_unreachable_case(Coordinate::new(4, 3));
        board[5][2] = case::create_unreachable_case(Coordinate::new(5, 2));
        board[5][3] = case::create_unreachable_case(Coordinate::new(5, 3));
        board[4][6] = case::create_unreachable_case(Coordinate::new(4, 6));
        board[4][7] = case::create_unreachable_case(Coordinate::new(4, 7));
        board[5][6] = case::create_unreachable_case(Coordinate::new(5, 6));
        board[5][7] = case::create_unreachable_case(Coordinate::new(5, 7));

        return board;
    }
}
